### PR TITLE
Nmoadev.contract logic

### PR DESCRIPTION
**Pull request by nmoadev**
* track a history of projects completed to satisfy some contracts
* add set of functions for general contract satisfaction
* fix a bug where workers weren't refunded when project completed

### CODE REVIEW

File: src/game_logic/turn.ts [hunk 1]
```diff
@@ -114,13 +118,17 @@ export function resolveContracts (initialState: GameState): GameState {
  * been satisfied or, if none was found, -1.
  */
 function findIndexOfSatisfiedOpenContract (state: GameState): number {
-  return state.player.contracts.open.findIndex(contract => contract.isSatisfied(state))
+  return state.player.contracts.open.findIndex(contract => {
+    const satisfied = contract.isSatisfied(state)
+    return satisfied
+  })
```
Comment (hypevhs) [inline src/game_logic/turn.ts]: Why? lol

**Review by scottysseus — APPROVED** [review]
noice

File: src/game_logic/contracts.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+/**
+ * Determine if the given gameState satisfies contains an object of desired completed project counts
+ * @param gameState
+ * @param desiredProjects
+ * @param comparator, optional argument to change how desired and actual are compared
+ * @returns true if for all properties of {desiredProjects}, at least the specified number of tiles has that tag
+ */
+export function satisifiesCompletedProjectCount (gameState: GameState, desiredProjects: DesiredCounts, comparator: Comparator = COMP_GTE): boolean {
```
Comment (hypevhs) [inline src/game_logic/contracts.ts]: typo

File: src/game_logic/interfaces.ts [hunk 3]
```diff
@@ -142,6 +142,9 @@ export interface PlayerState {
   readonly contracts: {
     readonly completed: ReadonlyArray<Contract>,
     readonly open: ReadonlyArray<Contract>
+  },
+  readonly projects: {
+    readonly completed: ProjectType[]
```
Comment (hypevhs) [inline src/game_logic/interfaces.ts]: use ReadonlyArray.

File: src/game_logic/shared.ts [hunk 4]
```diff
@@ -1,3 +1,14 @@
+import { TileCatalog } from '.'
```
Comment (hypevhs) [inline src/game_logic/shared.ts]: LINT!

File: src/game_logic/contracts.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+      satisfied = satisfied && comparator(actual, desired)
+    } else {
+      satisfied = false
+      break
+    }
+  }
+  return satisfied
+}
+
+type DesiredCounts = Record<string, number>
```
Comment (hypevhs) [inline src/game_logic/contracts.ts:61]: `string` is a little bleh, but even if we used WhateverType here TS wouldn't differentiate between them. Sad

File: src/game_logic/contracts.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+function countTilesByType (tiles: readonly Tile[]): Map<TileType, number> {
+  const result = new Map()
+  for (const tile of tiles) {
+    mapIncr(result, tile.type)
+  }
+  return result
+}
+
+function compareCounts<K extends string> (desiredCounts: Record<K, number>, actualCounts: Map<K, number>, comparator: Comparator): boolean {
```
Comment (hypevhs) [inline src/game_logic/contracts.ts]: Why not use DesiredCounts here too?

File: src/components/view/MapLocation.tsx [hunk 7]
```diff
@@ -22,7 +22,6 @@ export function MapLocation (props: {row: number, column: number, gridInterval:
   const onPointerOut = useCallback(() => setHover(false), [])
 
   const onClick = useCallback(() => {
-    console.log('Clicked on Map Location', tileDefinition.name, props.column, props.row)
     props.onSelected()
   }, [props, tileDefinition.name])
```
Comment (hypevhs) [inline src/components/view/MapLocation.tsx]: LINT!

File: src/game_logic/contracts.ts [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+import { GameState, ProjectType, Tile, TileType } from '.'
+import { tileCatalog } from '../data/tile-catalog'
+import { mapIncr } from './shared'
+
+type Comparator = (actual: number, desired: number) => boolean
+export const COMP_GTE: Comparator = (x, y) => x >= y
+export const COMP_EQ: Comparator = (x, y) => x === y
+
+function countTilesByTags (tiles: readonly Tile[]): Map<string, number> {
+  const result = new Map()
```
Comment (hypevhs) [inline src/game_logic/contracts.ts:14]: why not supply generic params?

File: src/game_logic/turn.ts (hunk 1 above)
Comment (nmoadev) [reply src/game_logic/turn.ts]: for debugging. Needed an easy line to place a breakpoint. will revert

**Pull request merged**